pandas/tests/indexes/interval/test_construction.py

```diff
@@ -111,6 +112,22 @@ def test_constructor_string(self, constructor, breaks):
         with tm.assert_raises_regex(TypeError, msg):
             constructor(**self.get_kwargs_from_breaks(breaks))
 
+    @pytest.mark.parametrize('cat_constructor', [
+        Categorical, CategoricalIndex])
+    def test_constructor_categorical_valid(self, constructor, cat_constructor):
+        # GH 21243/21253
+        if isinstance(constructor, partial) and constructor.func is Index:
+            # Index is defined to create CategoricalIndex from categorical data
+            pytest.skip()
```

Comment:
This is being skipped due to the following behavior:
```python
In [2]: cat = pd.Categorical([pd.Interval(0, 1), pd.Interval(1, 2), pd.Interval(0, 1)])

In [3]: pd.Index(cat, dtype='interval')
Out[3]: CategoricalIndex([(0, 1], (1, 2], (0, 1]], categories=[(0, 1], (1, 2]], ordered=False, dtype='category')
```

This happens because the `Index` code is structured so that categorical takes precedence over interval:
https://github.com/pandas-dev/pandas/blob/c85ab083919b59ce84c220d5baf7d34ff4a0bcf2/pandas/core/indexes/base.py#L262-L273

The code above could be restructured so that the `dtype` argument, if present, takes precedence over the type of `data`.  Seems like that would be more sensible than the current approach for this corner case, but on the fence about it.

Reply:
I c. ok can open an issue about this, but yes I would agree should infer with a passed dtype *first* before switching on the type of the data.